ksp_plugin/planetarium.cpp

```diff
@@ -117,10 +119,113 @@ RP2Lines<Length, Camera> Planetarium::PlotMethod1(
     }
     new_rp2_lines.push_back(std::move(new_rp2_line));
   }
-  LOG(INFO) << "Skipped " << skipped << " points out of " << total;
+  LOG(INFO) << "PlotMethod1 skipped " << skipped << " points out of " << total
+            << ".";
   return new_rp2_lines;
 }
 
+RP2Lines<Length, Camera> Planetarium::PlotMethod2(
+    DiscreteTrajectory<Barycentric>::Iterator const& begin,
+    DiscreteTrajectory<Barycentric>::Iterator const& end,
+    Instant const& now) const {
+  RP2Lines<Length, Camera> lines;
+  if (begin == end) {
+    return lines;
+  }
+  auto last = end;
+  --last;
+  if (begin == last) {
+    return lines;
+  }
+
+  auto const plottable_spheres = ComputePlottableSpheres(now);
+
+  auto const final_time = last.time();
+  auto const& trajectory = *begin.trajectory();
+
+  double const tan²_angular_resolution =
+      Pow<2>(parameters_.tan_angular_resolution_);
+
+  auto previous_time = begin.time();
+  auto previous_degrees_of_freedom = begin.degrees_of_freedom();
+
+  Position<Navigation> previous_position;
+  Velocity<Navigation> previous_velocity;
+  Time Δt;
+
+  previous_position =
+      plotting_frame_->ToThisFrameAtTime(previous_time)
+          .rigid_transformation()(previous_degrees_of_freedom.position());
+  previous_velocity =
+      plotting_frame_->ToThisFrameAtTime(previous_time)
+          .orthogonal_map()(previous_degrees_of_freedom.velocity());
+
+  Δt = final_time - previous_time;
+
+  Instant t;
+  double estimated_tan²_error;
+  Position<Navigation> position;
+
+  std::experimental::optional<Position<Navigation>> last_endpoint;
+
+  int steps_accepted = 0;
+  int steps_attempted = 0;
+
+  goto estimate_tan²_error;
+
+  while (previous_time < final_time) {
+    do {
+      Δt *= 0.9 * Sqrt(Sqrt(tan²_angular_resolution / estimated_tan²_error));
+    estimate_tan²_error:
+      t = previous_time + Δt;
+      if (t > final_time) {
+        t = final_time;
+        Δt = t - previous_time;
+      }
+      Position<Navigation> const extrapolated_position =
+          previous_position + previous_velocity * Δt;
+
+      position = plotting_frame_->ToThisFrameAtTime(t).rigid_transformation()(
+                     trajectory.EvaluatePosition(t));
+
+      estimated_tan²_error =
+          perspective_.Tan²AngularDistance(extrapolated_position, position) /
+          16;
+      ++steps_attempted;
+    } while (estimated_tan²_error > tan²_angular_resolution);
+    ++steps_accepted;
+
+    // TODO(egg): also limit to field of view.
+    auto const segment_behind_focal_plane =
```

Comment:
I cannot, I am overwriting previous_position (and there is the `continue` in between so I can't move that either).